pkg/utils/revision/revision_utils.go

```diff
@@ -0,0 +1,301 @@
+package revision
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"hash/fnv"
+
+	"github.com/davecgh/go-spew/spew"
+	appsv1 "k8s.io/api/apps/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	leaderworkerset "sigs.k8s.io/lws/api/leaderworkerset/v1"
+)
+
+// Functions in this package are adapted from https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/ and
+// https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/history/controller_history.go
+
+// ControllerRevisionName returns the Name for a ControllerRevision in the form prefix-hash-revisionnumber. If the length
+// of prefix is greater than 223 bytes, it is truncated to allow for a name that is no larger than 253 bytes.
+// revision-number allows us to avoid collisions if the created prefix-hash already exists in the history, since revision
+// will be unique.
+func RevisionName(prefix string, hash string, revisionNumber int64) string {
+	if len(prefix) > 223 {
+		prefix = prefix[:223]
+	}
+
+	return fmt.Sprintf("%s-%s-%v", prefix, hash, revisionNumber)
+}
+
+// HashControllerRevision hashes the contents of revision's Data using FNV hashing.
+// The returned hash will be a safe encoded string to avoid bad words.
+func HashRevision(revision *appsv1.ControllerRevision) string {
```

Comment:
can we move this and the functions related to hashing to the bottom beside each other and make them private (start with lower case ) please; I believe those functions should not be used outside the pkg